src/univariate/discrete/discretenormal.jl

```diff
@@ -0,0 +1,250 @@
+
+function sampler(rng::AbstractRNG, d::DiscreteNormal)
+    # Use a polyalgorithm. The tradeoff is a bit tricky because the primary cost
+    # of the alias table algorithm is the memory requirement, which Karney's
+    # algorithm doesn't have.
+    if d.σ > 100 # TODO: Determine the right value experimentally
+        return DiscreteNormalKarney(d)
+    else
+        return DiscreteNormalGeneric(d, AliasTable)
+    end
+end
+
+
+"""
+    An implementation of Karney's algorithm for sampling discrete gaussians
+    from [CFFK13], which modifications for floating point numbers from [MW19].
+
+    [CFFK13] Charles F. F. Karney "Sampling exactly from the normal distribution".
+           https://arxiv.org/abs/1303.6257
+"""
+module Karney
+    using Distributions
+    using Distributions: DiscreteNormal
+    using Random
+    import Random: rand
+
+    export DiscreteNormalKarney
+
+    struct DiscreteNormalKarney{T,S} <: Sampleable{Univariate,Discrete}
+        d::DiscreteNormal{T,S}
+    end
+
+    const exphalf = exp(-1/2)
+
+    function sample_ksj(rng, d)
+        # Step D1
+        k = (n = 0; while rand(rng, Bernoulli(exphalf)); n += 1; end; n)
+
+        # Step D2.
+        # TODO: Other implementations perform this computation using
+        # repeated Bernoulli trials of exp(-1/2). I am not sure why this is.
+        # Accuracy wise, we should get correctly rounded values over the
+        # entire support of the discrete gaussian. Perhaps this rounding is
+        # still too large?
+        # Performance wise, a Bernoulli trial is about 10x as expensive as
+        if k >= 2 && !rand(rng, Bernoulli(exp(-1/2*k*(k-1))))
+            return nothing
+        end
+        positive = rand(rng, Bool)
+        j::Int64 = rand(rng, DiscreteUniform(0, ceil(Int64, d.σ) - 1))
+
+        # Step D6 (c.f. CheckRejectB from [MW19])
+        if !positive && iszero(d.μ) && k == 0 && j == 0
+            return nothing
+        end
+
+        return (k, positive, j)
+    end
+
+    # Generic Karnery implementation
+    function rand(rng::AbstractRNG, d::DiscreteNormalKarney)
+        while true
+            tup = sample_ksj(rng, d.d)
+            tup === nothing && continue
+            (k, positive, j) = tup
+            (σ, μ) = (d.d.σ, d.d.μ)
+
+            # Step D4
+            sμ = positive ? μ : -μ
+            di₀ = σ * k + sμ
+            i₀::Int64 = ceil(Int64, di₀)
+            x₀ = i₀ - di₀
+            x = (x₀ + j)/σ
+
+            # Step D5
+            if (!isinteger(σ) && x >= 1)
+                continue
+            end
+
+            @assert x !== 0 || k !== 0
+            # Step D7
+            if !rand(rng, Bernoulli(exp(-1/2*x*(2k+x))))
+                continue
+            end
+
+            # Step D8
+            i::Int64 = i₀ + j
+            positive || (i = -i)
+
+            i in d.d.tail_cut || continue
+
+            return i
+        end
+    end
+
+    %₁(f) = modf(f)[1]
+
+    function isx̄zero(σ, μ, k, positive)
```

Comment:
Could we replace `isx̄zero` with `is_xzero` or at least `isxzero`?

Reply:
This checks for whether the variable `x̄` later will be zero, I'd rather keep it consistent with the paper.